File: 3-longest-substring-without-repeating-characters/longest-substring-without-repeating-characters.py
class Solution(object):
    def lengthOfLongestSubstring(self, s):
        """
        :type s: str
        :rtype: int
        """
        seen = set()
        left = 0
        maxCount = 0

        for right in range(len(s)):
            while(s[right] in seen):
                seen.remove(s[left])
                left += 1
            
            seen.add(s[right])
            maxCount = max(maxCount, right - left + 1)
        
        return maxCount

File: 3-longest-substring-without-repeating-characters/longest-substring-without-repeating-characters.java
class Solution {
    public int lengthOfLongestSubstring(String s) {
        Set<Character> unique = new HashSet<>();
        int l = 0;
        int r = 0;
        int maxLen = 0;
        while(r < s.length()){
            while(l <= r && unique.contains(s.charAt(r))){
                unique.remove(s.charAt(l));
                l++;
            }

            if(!unique.contains(s.charAt(r))){
                unique.add(s.charAt(r));
                maxLen = Math.max(maxLen, r - l + 1);
            }
            
            r++;
        }
        return maxLen; 
    }
}